core/src/Cache/CacheFallbackOperations.cs

```diff
@@ -238,29 +237,46 @@ internal class CacheFallbackOperations
         }
 
         public static AdalResultWrapper FindMsalEntryForAdal(ITokenCacheAccessor tokenCacheAccessor, string authority,
-            string clientId, string upn)
+            string clientId, string upn, RequestContext requestContext)
         {
-            foreach (var rtString in tokenCacheAccessor.GetAllRefreshTokensAsString())
-            {
-                var rtCacheItem =
-                    JsonHelper.DeserializeFromJson<MsalRefreshTokenCacheItem>(rtString);
+            var environment = new Uri(authority).Host;
 
-                var accountStr = tokenCacheAccessor.GetAccount(rtCacheItem.GetAccountItemKey());
-
-                var accountItem =
-                    JsonHelper.DeserializeFromJson<MsalAccountCacheItem>(accountStr);
+            List<MsalAccountCacheItem> accounts = new List<MsalAccountCacheItem>();
+            foreach (string accountStr in tokenCacheAccessor.GetAllAccountsAsString())
+            {
+                var accountItem = JsonHelper.TryToDeserializeFromJson<MsalAccountCacheItem>(accountStr, requestContext);
+                if (accountItem != null && accountItem.Environment.Equals(environment))
+                {
+                    accounts.Add(accountItem);
+                }
+            }
+            if (accounts.Count > 0)
+            {
+                foreach (var rtString in tokenCacheAccessor.GetAllRefreshTokensAsString())
+                {
+                    var rtCacheItem =
+                        JsonHelper.TryToDeserializeFromJson<MsalRefreshTokenCacheItem>(rtString, requestContext);
 
-                //TODO - authority check needs to be updated for alias check
-                if (new Uri(authority).Host.Equals(rtCacheItem.Environment) && rtCacheItem.ClientId.Equals(clientId) &&
-                    accountItem.PreferredUsername.Equals(upn))
-                    return new AdalResultWrapper
+                    //TODO - authority check needs to be updated for alias check
+                    if (rtCacheItem != null && environment.Equals(rtCacheItem.Environment)
```

Comment:
this is not relevant to this pr, in this code base Culture is not used for string comparison,
Please consider creating task

Reply:
I think this is for FxCop globalization checks in VSTS.

Reply:
This will break the compliance build if you try to check it in.

Reply:
IMO: should be addressed. Consider putting the comparisons into helper objects or into the objects them selves.

---
In reply to: [199528488](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1108#discussion_r199528488) [](ancestors = 199528488)

Reply:
Not sure about this one because it will break the compliance build. You can either disable this particular rule (it's FxCop based) and check that all these string comparisons methods w/o culture info do the right thing. Or you could be explicit and add culture information, which is what Wilson code does for example.

Reply:
I wonder how the build is green if the rules should stop it. Then we are possibly forcing the wrong build on the PR's

Reply:
@henrik-me - the build is green because we do not run all compliance tasks on PR build. I have a task to move some of these tasks there, but note that some of the build errors are not actionable by external contributors because the tools are internal.